photon-ml/src/main/scala/com/linkedin/photon/ml/data/ObjectProvider.scala

```diff
@@ -1,42 +0,0 @@
-/*
- * Copyright 2016 LinkedIn Corp. All rights reserved.
- * Licensed under the Apache License, Version 2.0 (the "License"); you may
- * not use this file except in compliance with the License. You may obtain a
- * copy of the License at
- *
- * http://www.apache.org/licenses/LICENSE-2.0
- *
- * Unless required by applicable law or agreed to in writing, software
- * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
- * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
- * License for the specific language governing permissions and limitations
- * under the License.
- */
-package com.linkedin.photon.ml.data
-
-
-import org.apache.spark.broadcast.Broadcast
-
-
-/**
- * A wrapper for normal object and broadcast object. This approach simplifies logic for classes/methods that deal with
- * both iterable and RDD data.
- *
- * This trait hides the logic whether the data are local or from remote. Probably it will be better if Iterable and
- * RDD data are separated out in [[com.linkedin.photon.ml.function.DiffFunction]] and
- * [[com.linkedin.photon.ml.function.TwiceDiffFunction]] classes.
- *
- */
-trait ObjectProvider[T <: Serializable] extends Serializable {
-  def get: T
-}
-
-@SerialVersionUID(1L)
-class SimpleObjectProvider[T <: Serializable](obj: T) extends ObjectProvider[T] {
-  override def get: T = obj
-}
-
-@SerialVersionUID(1L)
-class BroadcastedObjectProvider[T <: Serializable](obj: Broadcast[T]) extends ObjectProvider[T] {
-  override def get: T = obj.value
-}
```

Comment:
The two code paths were originally overlapping because in the end the code they want to call is the same: it's the code in the `ValueAndGradientAggregator` and `HessianVectorAggregator`. For fixed effect problems, we need to perform a distributed aggregation. For random effect problems, it is a localized one.

However, to merge the logic paths in one class meant that the class either had to have multiple identical methods (one for each input data type) or it had to take input objects of multiple data types in a wrapper object. Currently it does the latter; some objects using the `ObjectProvider`, and some using `Either`.

In addition, the logic of the two problems is not completely identical. The fixed effect problem should **have** an RDD of data, should not be serializable, and should perform down-sampling. The individual random effect problems should **be** an RDD of optimization problems that handle data locally and be serializable.

It is my opinion that these differences are substantial enough (the serialization relationship alone) that they should be made explicit in the structure of the code.